libroach: Migrate IncrementalIterator logic to C++

adityamaru27 opened the pull request:
Export was previously executed using an `MVCCIncrementalIterator`
where the logic handling iteration over the diff of the key
range `[startKey, endKey)` and time range `(startTime, endTime]`
was written in Go. This iteration involved making a cgo call
to find every key, along with another cgo call for writing
each key to the sstable.
This migration resolves the aforementioned performance
bottleneck (#18884), by pushing all the required logic to C++
and exposing a single export method.

Based on a performance benchmark by running BACKUP on a tpcc database with 1000 warehouses we observe the following:

- Over an average of 3 runs we see a 1.1x improvement in time performance. While the original binary took ~32m04s the changed implementation took ~28m55s. This is due to the elimination of a cgo call per key.

adityamaru27 committed:
engine: Add ExportToSst method with all logic moved to C++

Export was previously executed using an MVCCIncrementalIterator
where the logic handling iteration over the diff of the key
range [startKey, endKey) and time range (startTime, endTime]
was written in Go. This iteration involved making a cgo call
to find every key, along with another cgo call for writing
each key to the sstable.
This migration resolves the aforementioned performance
bottleneck (#18884), by pushing all the required logic to C++
and exposing a single export method.

Release note: Performance improvement to the BACKUP process.

adityamaru27 committed:
libroach: Migrate MVCCIncrementalIterator logic to C++

All of the IncrementalIterator logic was previously written in Go.
This change migrates all the required implementations to C++ and
allows the DBExportToSst method to use this iterator to export
the required keys to an SSTable.

Release note: Performance improvement to the BACKUP process.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37496) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37496) before we can accept your contribution.<br/><hr/>**Aditya Maru** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=37496) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37496)
<!-- Reviewable:end -->

petermattis commented:
Also tagging @ajkr on this as he's familiar with both C++ and RocksDB.

dt commented:
@ajkr you can hold off for a bit, we've discussed this a bit with @danhhz and are going to change the approach slightly and revise this before it is ready for review.

dt commented on pkg/ccl/storageccl/export.go:
```diff
@@ -91,85 +90,69 @@ func evalExport(
-	var rows bulk.RowCounter
-	// TODO(dan): Move all this iteration into cpp to avoid the cgo calls.
-	// TODO(dan): Consider checking ctx periodically during the MVCCIterate call.
-	iter := engineccl.NewMVCCIncrementalIterator(batch, engineccl.IterOptions{
```
Suggestion: you could pull this code into a go function that takes the same args are your new c++ function (and keep mvcc.go around for now). Then you could write a unit test that calls both functions on the same engine and compares the result?

adityamaru27 replied:
Done.

adityamaru27 replied:
Done.

dt reviewed (commented):
> Reviewed 3 of 3 files at r1, 12 of 14 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37496)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27 and @ajkr)

___
*[c-deps/libroach/db.h, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexDA9O8oBFrEqfZmM6:-LexDA9O8oBFrEqfZmM7:b5t05jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/db.h#L104)):*
> ```objc
> // preferred function for Go callers.
> DBStatus DBSstFileWriterAddRaw(DBSstFileWriter* fw, const ::rocksdb::Slice key,
>                                const ::rocksdb::Slice val);
> ```

nit: newline at end of file
___
*[c-deps/libroach/db.cc, line 950 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexI9q37TyheCKDUFtq:-LexI9q465CplO_4rpyS:b-lqcuk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/db.cc#L950)):*
> ```cpp
>     // Skip tombstone (len=0) records when start time is zero (non-incremental)
>     // and we are not exporting all versions.
>     if (!export_all_revisions && iter.value().size() == 0 && start.wall_time == 0 &&
> ```

2¢: I might pull the three conditions that go into "are we skipping deletes?" into a bool before the loop, so we only check "are we skipping deletes && is this a delete"
___
*[c-deps/libroach/db.cc, line 974 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexFjTk9wwIpb3CQDj1:-LexFjTk9wwIpb3CQDj2:b5t05jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/db.cc#L974)):*
> ```cpp
> 
>   return res;
> }
> ```

nit: newline at end of file
___
*[c-deps/libroach/incremental_iterator.h, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexEtb-3t59KFyQeErg:-LexEtb-3t59KFyQeErh:bxqvv28) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.h#L1)):*
> ```objc
> // Copyright 2018 The Cockroach Authors.
> ```

nit: 2019
___
*[c-deps/libroach/incremental_iterator.h, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexE1r56i_WnAdMvq3B:-LexE1r56i_WnAdMvq3C:b5t05jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.h#L52)):*
> ```objc
>   cockroach::util::hlc::LegacyTimestamp start_time;
>   cockroach::util::hlc::LegacyTimestamp end_time;
> };
> ```

nit: newline at end of file
___
*[c-deps/libroach/incremental_iterator.cc, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexEoMtEzB7W84MB1um:-LexEoMu5EgiJjOnxRNG:bxqvv28) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.cc#L1)):*
> ```cpp
> // Copyright 2018 The Cockroach Authors.
> ```

nit: 2019
___
*[c-deps/libroach/incremental_iterator.cc, line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexETtX8oRWj6LpWEiq:-LexETtX8oRWj6LpWEir:b5t05jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.cc#L223)):*
> ```cpp
> const rocksdb::Slice DBIncrementalIterator::key() { return iter.get()->rep->key(); }
> 
> const rocksdb::Slice DBIncrementalIterator::value() { return iter.get()->rep->value(); }
> ```

nit: newline at end of file
___
*[pkg/storage/engine/rocksdb.go, line 3201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexCGaX7zm4nBvdnAmH:-LexCGaX7zm4nBvdnAmI:b-w23o38) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3de2b0907fa34ab4efca7a9fb42e83fc570b7bf/pkg/storage/engine/rocksdb.go#L3201)):*
> ```Go
> func ExportToSst(
> 	ctx context.Context, e Reader, start, end MVCCKey, exportAllRevisions bool, io IterOptions,
> ) ([]byte, C.int64_t, error) {
> ```

I'd return an `int64` instead of the c type.
___
*[pkg/storage/engine/rocksdb.go, line 3213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexCut43dm4iOV_ZasM:-LexCut43dm4iOV_ZasN:bvt5coj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3de2b0907fa34ab4efca7a9fb42e83fc570b7bf/pkg/storage/engine/rocksdb.go#L3213)):*
> ```Go
> 	var intentErr C.DBString
> 
> 	err := statusToError(C.DBExportToSst(goToCKey(start), goToCKey(end), C.bool(exportAllRevisions),
> ```

I think you might have inverted the order of the commits?
___
*[pkg/storage/engine/rocksdb.go, line 3223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexCL3i-YbgxTiHOkSm:-LexCL3i-YbgxTiHOkSn:b9zvtw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3de2b0907fa34ab4efca7a9fb42e83fc570b7bf/pkg/storage/engine/rocksdb.go#L3223)):*
> ```Go
> 			}
> 
> 			return nil, dataSize, &e
> ```

I'd return 0 instead of the undefined value (right?) in dataSize.


<!-- Sent from Reviewable.io -->

adityamaru27 committed:
spanset: Exported SpanSetReadWriter to be used in export logic.

The export logic requires access to the C.DBEngine of an
engine.Reader. In the case of a spanSetReader this
requires us to recursively access the engine.Reader member
until we reach either a RocksDB or rocksDBReadOnly engine.
A type switch construct is deemed necessary in
storageccl/export.go, requiring the SpanSetReadWriter type to
be exported.

Note: lint fails if we name it SpanSetReadWriter post exporting,
as other packages would have to access it using
spanset.SpanSetReadWriter which introduces a "stutter". It has
been renamed to ReadWriter.

Release note: None

adityamaru27 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37496)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @dt)

___
*[c-deps/libroach/db.cc, line 950 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexI9q37TyheCKDUFtq:-Lf1h_Av0-jgR2AM3O_K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/db.cc#L950)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

2¢: I might pull the three conditions that go into "are we skipping deletes?" into a bool before the loop, so we only check "are we skipping deletes && is this a delete"
</blockquote></details>

Done.
___
*[c-deps/libroach/db.cc, line 974 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexFjTk9wwIpb3CQDj1:-Lf1hbOl0z5GOagfrggL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/db.cc#L974)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: newline at end of file
</blockquote></details>

Done.
___
*[c-deps/libroach/incremental_iterator.h, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexEtb-3t59KFyQeErg:-Lf1hlcL20B8FMzbskq-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.h#L1)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: 2019
</blockquote></details>

Done.
___
*[c-deps/libroach/incremental_iterator.h, line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexE1r56i_WnAdMvq3B:-Lf1hnOf2T7UmJe2RuJK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.h#L52)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: newline at end of file
</blockquote></details>

Done.
___
*[c-deps/libroach/incremental_iterator.cc, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexEoMtEzB7W84MB1um:-Lf1hpKr8yvkZwku6-VJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.cc#L1)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: 2019
</blockquote></details>

Done.
___
*[c-deps/libroach/incremental_iterator.cc, line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexETtX8oRWj6LpWEiq:-Lf1hq2l0UJKcmWc-174:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f508ca222a479868e54be7ed22648fc473934ea4/c-deps/libroach/incremental_iterator.cc#L223)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

nit: newline at end of file
</blockquote></details>

Done.
___
*[pkg/storage/engine/rocksdb.go, line 3201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexCGaX7zm4nBvdnAmH:-Lf1hxiw5hd7yd4AtQcC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3de2b0907fa34ab4efca7a9fb42e83fc570b7bf/pkg/storage/engine/rocksdb.go#L3201)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'd return an `int64` instead of the c type.
</blockquote></details>

Done.
___
*[pkg/storage/engine/rocksdb.go, line 3213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexCut43dm4iOV_ZasM:-Lf1j8eE9jkOu7wWwQVb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3de2b0907fa34ab4efca7a9fb42e83fc570b7bf/pkg/storage/engine/rocksdb.go#L3213)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think you might have inverted the order of the commits?
</blockquote></details>

Done.
___
*[pkg/storage/engine/rocksdb.go, line 3223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LexCL3i-YbgxTiHOkSm:-Lf1i0TGE_HyTgSl36bl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3de2b0907fa34ab4efca7a9fb42e83fc570b7bf/pkg/storage/engine/rocksdb.go#L3223)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'd return 0 instead of the undefined value (right?) in dataSize.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adityamaru27 committed:
storageccl: Added test to check export correctness with random input.

Added a test which generates random KVPs and timestamps, subsequently
attempting to export the data between provided key and time ranges.
Correctness of the new export logic is checked against the old
implementation which used the now non-existant MVCCIncrementalIterator.

Release note: None

dt commented:
@ajkr thanks for waiting -- this is now ready for review now!

ajkr commented:
Thanks, taking a look. Are there any benchmark results on export time/CPU with vs without this PR?

adityamaru27 commented:
> Thanks, taking a look. Are there any benchmark results on export time/CPU with vs without this PR?

Hey @ajkr! Not yet, but I am working on compiling some results from roachprod and existing export/backup benchmarks. I shall update the PR as soon as I have compiled them.

ajkr reviewed (commented):
LGTM, all my comments are minor.

A couple high-level questions.
- why does `ExportToSst` return a byte array rather than take a filename as an argument and export to that file?
- what limits the size of the SST created by `ExportToSst`?

> Reviewed 1 of 18 files at r3, 14 of 14 files at r4, 4 of 4 files at r5, 1 of 1 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-:-LfRd-E9DEIC5ex7Bfd6:b-berfp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27 and @dt)

___
*[c-deps/libroach/db.h, line 103 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRKtDR5ICEomPoMERx:-LfRKtDR5ICEomPoMERy:b-1wa21q) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/db.h#L103)):*
> ```objc
> // DBSstFileWriterAddRaw is used internally -- DBSstFileWriterAdd is the
> // preferred function for Go callers.
> DBStatus DBSstFileWriterAddRaw(DBSstFileWriter* fw, const ::rocksdb::Slice key,
> ```

Do we need this declaration in `db.h`? I think it's only used in `db.cc`, so we should be able to declare and define it in an anonymous namespace there. (Similar to `DBIterGetState()`, for example.)
___
*[c-deps/libroach/db.cc, line 939 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRPOPz7vdgzW9vgPW5:-LfRPOPz7vdgzW9vgPW6:bwmkkzc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/db.cc#L939)):*
> ```cpp
>       DBSstFileWriterClose(writer);
>       return state.status;
>     } else if (!state.valid || kComparator.Compare(iter.key(), EncodeKey(end)) >= 0) {
> ```

We could avoid an allocation/memcpy on each iteration by getting the result of `EncodeKey(end)` before the for-loop. I don't know whether it'll make a noticeable difference, but typically I have seen memcpy/malloc/free near the top of CPU profiles for other range scan workloads.
___
*[c-deps/libroach/incremental_iterator.h, line 28 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRReHTD3Nd5Kad_OqA:-LfRReHTD3Nd5Kad_OqB:b-9d2qpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/incremental_iterator.h#L28)):*
> ```objc
>                         DBString* write_intent);
>   ~DBIncrementalIterator();
>   void advanceKey();
> ```

can `advanceKey()` be private?
___
*[c-deps/libroach/incremental_iterator.cc, line 52 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRTeSR1yxgAaaj75K9:-LfRTeSR1yxgAaaj75KA:b4ozwi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/incremental_iterator.cc#L52)):*
> ```cpp
>     assert(!EmptyTimestamp(opts.max_timestamp_hint));
>     DBIterOptions nontimebound_opts = DBIterOptions();
>     nontimebound_opts.upper_bound = opts.upper_bound;
> ```

I'm kind of curious about the relationship between `opts.upper_bound` and `this->end`. Will they be the same?
___
*[pkg/ccl/storageccl/export.go, line 164 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRbGZoAIAEMSEL0oN4:-LfRbGZp3z_5BOHCDhZ6:bprfxat) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/pkg/ccl/storageccl/export.go#L164)):*
> ```Go
> 		}
> 
> 		rows.BulkOpSummary.DataSize += int64(len(it.UnsafeKey().Key)) + int64(len(it.UnsafeValue()))
> ```

Is the result being computed here different from what's stored in the variable `dataSize`?


<!-- Sent from Reviewable.io -->

dt commented:
> why does ExportToSst return a byte array rather than take a filename as an argument and export to that file?

In some cases we return the content in the response rather than write a file at all, and even when we're writing a "file" it is usually to remote, cloud storage, not the local file system. The cloud storage accessors are in Go, so passing the content to write back up and letting go decide what to do with it seems to make sense.

FWIW, prior to this change, we were also passing the same slice back from c++ to go when we called `Finish` on the `RocksDBSstFileWriter`.

> what limits the size of the SST created by ExportToSst?

Range size.

adityamaru27 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37496)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajkr and @dt)

___
*[c-deps/libroach/db.h, line 103 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRKtDR5ICEomPoMERx:-LfV-bNkDJS82lwenByo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/db.h#L103)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

Do we need this declaration in `db.h`? I think it's only used in `db.cc`, so we should be able to declare and define it in an anonymous namespace there. (Similar to `DBIterGetState()`, for example.)
</blockquote></details>

Done.
___
*[c-deps/libroach/db.cc, line 939 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRPOPz7vdgzW9vgPW5:-LfV-e-F1yGNccSgI_zI:boj7l8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/db.cc#L939)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

We could avoid an allocation/memcpy on each iteration by getting the result of `EncodeKey(end)` before the for-loop. I don't know whether it'll make a noticeable difference, but typically I have seen memcpy/malloc/free near the top of CPU profiles for other range scan workloads.
</blockquote></details>

makes sense, done.
___
*[c-deps/libroach/incremental_iterator.h, line 28 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRReHTD3Nd5Kad_OqA:-LfV-iRDBldTKRYPQqM8:b4j0me7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/incremental_iterator.h#L28)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

can `advanceKey()` be private?
</blockquote></details>

yes it can, done.
___
*[c-deps/libroach/incremental_iterator.cc, line 52 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRTeSR1yxgAaaj75K9:-LfV-EGE12IfH4Ikwa59:b-ahi06x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/c-deps/libroach/incremental_iterator.cc#L52)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

I'm kind of curious about the relationship between `opts.upper_bound` and `this->end`. Will they be the same?
</blockquote></details>

`opts.upper_bound` is a DBKey with empty `wall_time` and `logical` fields (i.e. it only enforces a key upper bound). `this->end` has those fields populated, and is used in `advanceKey()` for the purpose of checking that we are within the required time bounds.
___
*[pkg/ccl/storageccl/export.go, line 164 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRbGZoAIAEMSEL0oN4:-LfV2_m32QZ1iLKwlUNW:bw0vn4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/pkg/ccl/storageccl/export.go#L164)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

Is the result being computed here different from what's stored in the variable `dataSize`?
</blockquote></details>

So using dataSize instead of recomputing it here was causing existing backup tests to fail, as the computed bytes were more than the expected bytes. I initially thought this could be because we are iterating more keys than we should on the C++ side, but the number of rows and the SSTable data itself matched up to what was expected. 
I use dataSize to ensure that we are in fact exporting some data (otherwise the levelDB iterator collecting stats after the cgo call panics) and then recompute the actual `rows.BulkOpSummary.DataSize` as was the case in the previous implementation.
Eventually, we would like to move the stats collection (including RowCounter) to C++, but decided it was not in the scope of this change.


<!-- Sent from Reviewable.io -->

ajkr reviewed (approved):
> Reviewed 1 of 20 files at r9, 18 of 19 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37496)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27 and @dt)

___
*[c-deps/libroach/db.cc, line 937 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfVZX8MDG6qBfEf83Sd:-LfVZX8NCg_PtG-qwpzI:bytnsby) ([raw file](https://github.com/cockroachdb/cockroach/blob/24765a1919ed347fe5fe3a4559a57edfce26fe7b/c-deps/libroach/db.cc#L937)):*
> ```cpp
>   bool skip_current_key_versions = !export_all_revisions;
>   DBIterState state;
>   const rocksdb::Slice end_key = EncodeKey(end);
> ```

I think we need to store the string result otherwise the memory pointed to by the `Slice` can be freed.
___
*[pkg/ccl/storageccl/export.go, line 164 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfRbGZoAIAEMSEL0oN4:-LfV__Sd9RItCepjFcNu:bx40ouy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d6628ec9e174f9aca327f0069eda4f2e98a77865/pkg/ccl/storageccl/export.go#L164)):*
<details><summary><i>Previously, adityamaru27 (Aditya Maru) wrote…</i></summary><blockquote>

So using dataSize instead of recomputing it here was causing existing backup tests to fail, as the computed bytes were more than the expected bytes. I initially thought this could be because we are iterating more keys than we should on the C++ side, but the number of rows and the SSTable data itself matched up to what was expected. 
I use dataSize to ensure that we are in fact exporting some data (otherwise the levelDB iterator collecting stats after the cgo call panics) and then recompute the actual `rows.BulkOpSummary.DataSize` as was the case in the previous implementation.
Eventually, we would like to move the stats collection (including RowCounter) to C++, but decided it was not in the scope of this change.
</blockquote></details>

Got it, thanks for the explanation.


<!-- Sent from Reviewable.io -->

adityamaru27 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37496)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru27, @ajkr, and @dt)

___
*[c-deps/libroach/db.cc, line 937 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/37496#-LfVZX8MDG6qBfEf83Sd:-LfVgsB3-dSm7drOwYLW:bpotwk9) ([raw file](https://github.com/cockroachdb/cockroach/blob/24765a1919ed347fe5fe3a4559a57edfce26fe7b/c-deps/libroach/db.cc#L937)):*
<details><summary><i>Previously, ajkr (Andrew Kryczka) wrote…</i></summary><blockquote>

I think we need to store the string result otherwise the memory pointed to by the `Slice` can be freed.
</blockquote></details>

Thanks for this catch! I was scratching my head as to why the tests were failing with these changes.


<!-- Sent from Reviewable.io -->

adityamaru27 commented:
@dt changed the headers of the newly added files to BSL. Will merge once you have a quick look!

adityamaru27 commented:
bors r=dt

adityamaru27 commented:
bors r=dt
